File: 2969-minimum-number-of-coins-for-fruits-ii/2969-minimum-number-of-coins-for-fruits-ii.java
class Solution {
    public int minimumCoins(int[] prices) {
        int n = prices.length;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        pq.offer(new int[] {2, prices[0]});
        for (int i = 1; i < n; i++) {
            while (!pq.isEmpty() && pq.peek()[0] < i)
                pq.poll();
            int[] cur = pq.peek();
            int p = cur[1];
            pq.offer(new int[] {i * 2 + 2, p + prices[i]});
        }
        while (pq.peek()[0] < n)
            pq.poll();
        return pq.peek()[1];
    }
}

File: 2969-minimum-number-of-coins-for-fruits-ii/2969-minimum-number-of-coins-for-fruits-ii.py
class Solution:
    def minimumCoins(self, prices: List[int]) -> int:
        n = len(prices)
        buy, not_buy = prices[0], inf
        heap = [(prices[0], 1)]
        for i in range(2, n + 1):
            buy = min(buy, not_buy) + prices[i - 1]
            t = ceil(i / 2)
            while heap[0][1] < t:
                heappop(heap)
            not_buy = heap[0][0]
            heappush(heap, (buy, i))
        return min(buy, not_buy)

